components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+
+/// Download a Habitat package.
+///
+/// If an `PackageIdentTarget` is given, we retrieve the package
+/// from the specified Builder `url`. Providing a fully-qualified
+/// identifer will result in that exact package being installed
+/// (regardless of `channel`). Providing a partially-qualified
+/// identifier will result in the installation of latest appropriate
+/// release from the given `channel`.
+///
+/// Any dependencies of will be retrieved from Builder (if they're not
+/// already cached locally).
+///
+/// At the end of this function, the specified package and all its
+/// dependencies will be downloaded on the system.
+
+/// Note: it's worth investigating whether
+/// LocalPackageUsage makes sense here
+/// Also, in the future we may want to accept an alternate builder to 'filter' what we pull down by
+/// That would greatly optimize the 'sync' to on prem builder case, as we could point to that
+/// and only fetch what we don't already have.
+#[allow(clippy::too_many_arguments)]
+pub fn start<U>(ui: &mut U,
+                url: &str,
+                channel: &ChannelIdent,
+                product: &str,
+                version: &str,
+                idents: Vec<PackageIdent>,
+                target: PackageTarget,
+                fs_root_path: Option<&PathBuf>,
+                token: Option<&str>)
+                -> Result<()>
+    where U: UIWriter
+{
+    debug!("Starting download with url: {}, channel: {}, product: {}, version: {}, target: {}, \
+            fs_root_path: {:?}, token: {:?}",
+           url, channel, product, version, target, fs_root_path, token);
+
+    let key_cache_path = &cache_key_path(fs_root_path);
+    debug!("install key_cache_path: {}", key_cache_path.display());
+
+    let artifact_cache_path = &cache_artifact_path(fs_root_path);
+    debug!("install artifact_cache_path: {}",
+           artifact_cache_path.display());
+
+    // TODO we use the same root path for ssl certs as we do for the rest of the root path,
+    // We shouldn't probably override it here, as this appears to be largely for cert paths
+    let api_client = Client::new(url, product, version, fs_root_path.map(PathBuf::as_path))?;
+    let task = DownloadTask { idents,
+                              target,
+                              url,
+                              api_client,
+                              token,
+                              channel,
+                              artifact_cache_path,
+                              key_cache_path };
+
+    let downloaded_artifacts: Vec<PackageArchive> = task.execute(ui).unwrap();
+
+    debug!("Expanded package count: {}", downloaded_artifacts.len());
+
+    Ok(())
+}
+
+struct DownloadTask<'a> {
+    idents: Vec<PackageIdent>,
+    target: PackageTarget,
+    url: &'a str,
+    api_client: BoxedClient,
+    token: Option<&'a str>,
+    channel: &'a ChannelIdent,
+    /// The path to the local artifact cache (e.g., /hab/cache/artifacts)
```

Comment:
Instead of calling these "cache" paths, maybe just call them "download paths"?  Also do we need to pass both artifact and key cache separately?